superset/assets/src/visualizations/deckgl/CategoricalDeckGLContainer.jsx

```diff
@@ -103,7 +103,9 @@ export default class CategoricalDeckGLContainer extends React.PureComponent {
       data = data.filter(d => this.state.categories[d.cat_color].enabled);
     }
 
-    return [this.props.getLayer(fd, data, this.props.slice)];
+    const payload = this.props.payload;
+    payload.data.features = data;
+    return [this.props.getLayer(fd, payload, this.props.slice)];
```

Comment:
This can simply be written as:

```javascript
return [this.props.getLayer(fd, this.props.payload, this.props.slice)];
```

If you follow my comment above. You can also do in the beginning of the function:

```javascript
const { getLayer, payload, slice } = this.props;
```

The you can simple refer to `getLayer`, `payload` and `slice` inside the function, instead of `this.props.getLayer`, `this.props.payload` and `this.props.slice`.

Reply:
A lot of good educations here! Thanks!
The reason why I am making a copy of `this.props.payload.data.features`, mutating it, and then assigning it back to payload in line 107 is that if I understand correctly, changing in `this.props` will result in re-render. As we are only passing it to `getLayer` function and don't need it to be re-rendered, I copy it so `props` is unaffected and unnecessary re-render can be avoided. This is my understanding but I am not very sure whether it is correct based on your suggestion. Please correct me if I am wrong.

Reply:
well. I am wrong. Only state change will cause re-render.